sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/dynamodb/DynamoDBIO.java

```diff
@@ -0,0 +1,533 @@
+      checkArgument(
+          (scanRequest.totalSegments() != null && scanRequest.totalSegments() > 0),
+          "TotalSegments is required with withScanRequestFn() and greater zero");
+
+      PCollection<Read<T>> splits =
+          (PCollection<Read<T>>)
+              input.apply("Create", Create.of(this)).apply("Split", ParDo.of(new SplitFn()));
+      splits.setCoder(SerializableCoder.of(new TypeDescriptor<Read<T>>() {}));
+
+      PCollection<T> output =
+          (PCollection<T>)
+              splits
+                  .apply("Reshuffle", Reshuffle.viaRandomKey())
+                  .apply("Read", ParDo.of(new ReadFn()));
+      output.setCoder(getCoder());
+      return output;
+    }
+
+    /** A {@link DoFn} to split {@link Read} elements by segment id. */
+    private static class SplitFn<T> extends DoFn<Read<T>, Read<T>> {
+      @ProcessElement
+      public void processElement(@Element Read<T> spec, OutputReceiver<Read<T>> out) {
+        ScanRequest scanRequest = spec.getScanRequestFn().apply(null);
+        for (int i = 0; i < scanRequest.totalSegments(); i++) {
+          out.output(spec.withSegmentId(i));
+        }
+      }
+    }
+
+    /** A {@link DoFn} executing the ScanRequest to read from DynamoDB. */
+    private static class ReadFn<T> extends DoFn<Read<T>, T> {
+      @ProcessElement
+      public void processElement(@Element Read<T> spec, OutputReceiver<T> out) {
+        DynamoDbClient client = spec.getAwsClientsProvider().createDynamoDB();
+
+        ScanRequest scanRequest = spec.getScanRequestFn().apply(null);
+        ScanRequest scanRequestWithSegment =
+            scanRequest.toBuilder().segment(spec.getSegmentId()).build();
+
+        ScanResponse scanResponse = client.scan(scanRequestWithSegment);
+        out.output(spec.getScanResponseMapperFn().apply(scanResponse));
+      }
+    }
+
+    static final class ItemsMapper<T>
+        implements SerializableFunction<ScanResponse, List<Map<String, AttributeValue>>> {
+      @Override
+      public List<Map<String, AttributeValue>> apply(@Nullable ScanResponse scanResponse) {
+        if (scanResponse == null) {
+          return Collections.emptyList();
+        }
+        return scanResponse.items();
+      }
+    }
+  }
+
+  /**
+   * A POJO encapsulating a configuration for retry behavior when issuing requests to dynamodb. A
+   * retry will be attempted until the maxAttempts or maxDuration is exceeded, whichever comes
+   * first, for any of the following exceptions:
+   *
+   * <ul>
+   *   <li>{@link IOException}
+   * </ul>
+   */
+  @AutoValue
+  public abstract static class RetryConfiguration implements Serializable {
+    @VisibleForTesting
+    static final RetryPredicate DEFAULT_RETRY_PREDICATE = new DefaultRetryPredicate();
+
+    abstract int getMaxAttempts();
+
+    abstract Duration getMaxDuration();
+
+    abstract RetryPredicate getRetryPredicate();
+
+    abstract Builder builder();
```

Comment:
Rename to `toBuilder()` and add a new method:
```
   public static Builder builder() {
    return new AutoValue_RetryConfiguration.Builder();
  }
```
This is to be consistent with the good practices of AutoValue Builder use (This is something we are not doing correctly in many IOs except the google ones). For more details https://github.com/google/auto/blob/master/value/userguide/builders-howto.md